sam.c

```diff
@@ -4955,6 +4955,164 @@ char *sam_open_mode_opts(const char *fn,
     return mode_opts;
 }
 
+// Parses one item, which may be a set of flags specified numerically
+// or via symbolic letter codes, or a single flag via its long name.
+static unsigned long sam_parse_flag1(const char *s, char **endptr)
+{
+    if (isdigit_c(*s)) return strtoul(s, endptr, 0);
+
+    unsigned long flag = 0;
+    char name[24];
+    int i = 0, short_ok = 1;
+
+    for (; isalpha_c(*s) || *s == '_'; s++) {
+        if (i < sizeof name) name[i++] = toupper_c(*s);
```

Comment:
Yeah, I have an irrational disliking for `strcasecmp` and friends. The claimed rational basis is that in this checking against a table of keywords scenario, using `strncasecmp` implies O(N²) locale-aware case conversions (really O(MN) where M is the number of entries in the table and N is the average word length) while canonicalising the input costs only O(N) case conversions.

But admittedly computers are fast nowadays and `strncasecmp` is indeed used elsewhere in htslib… :shrug:

I have a followup commit that rewrites `bam_str2flag` in terms of `sam_parse_flag1` but I didn't want to muddy the waters with it just yet :-).

Reply:
Re: `bam_str2flag`: we can do it in two stages fine.

Regarding `strncasecmp`, I don't follow the notion of higher complexity.  With a string limit it's just the same deal, and it can be used just as before in a switch statement to restrict the search space.  At least I'd always *assumed* the only difference between strncasecmp and strncmp is the initial case permutation of each character (or word) before comparing.  But anyway, as I say it's not a major concern.  I was just curious why the extra temporary buffer was used.

Reply:
If you do M `strncasecmp(input, keyword[i])` checks (for M valid keywords and with `0 <= i < M`), you have to canonicalise each character of `input` M times and each character of each of the M `keyword` strings once apiece — this calculation is complicated by only having to canonicalise the pairs of characters at the start of the strings that you actually need to compare, but is nonetheless O(MN).

OTOH if you write your `keyword` table pre-canonicalised, then you only need to canonicalise each character of `input` exactly once, which is O(N). In fact, it's exactly N character case conversions.

I said it was irrational! I may rewrite this using the combination of the existing `switch (firstchar)` style with `strncasecmp`. As you say, the extra fixed length buffer is not great either.

Reply:
Hmm, for M strings being compared of length N, I'd see both `strncasecmp` and `strcasecmp` being O(MN), but with different constant scaling factors.

Personally for printable strings I'd do something as trivial as `if ((c1 & ~0x20) != (c2 & ~0x20)) break` (case insensitive) vs `if (c1 != c2) break` (case sensitive), but looking at the code I see it's a bit more flexible and has a `charmap[u1] != charmap[u2]` lookup instead.  That basically means the canonicalisation is simply a constant slow down on any comparison.

Agreed it's perhaps slower than doing it once upfront to avoid the canonicalisation bit being repeated, but the complexity itself is the same order.  Anyway, it's arguing over something that frankly irrelevant.  You've got to be pretty special to have argument parsing being the bottleneck of your work. :-)  I don't really mind either way.